### PR TITLE
Adjust \underset alignment and spacing

### PR DESCRIPTION
**Pull request by ronkok**
Fixes issue #1288.

1. Fixes vertical alignment by keeping the base element on the baseline.
2. Assigns horizontal spacing to \underset in a manner that aligns with LaTeX `\binrel`.

### CODE REVIEW

**Comment by ronkok** [comment]
Results:
adjustunderset

**Comment by ronkok** [comment]
I'll be curious to look at all the other screenshotter tests. The change I made in `op.js` should not change anything but `\underset`. We'll see if that's true.

**Comment by edemaine** [comment]
I updated the `OverUnderset` screenshot.  Thanks for adding the wiggle test to avoid this breaking in the future!

**Comment by ronkok** [comment]
@edemaine The screenshot test looks good to me. Thank you.

File: src/functions.js [hunk 1]
```diff
@@ -66,26 +66,39 @@ defineFunction([
...
-    const xAxisOp = new ParseNode("op", {
+    let mclass = "mrel";
+    if (context.funcName !== "\\stackrel") {
+        // LaTeX applies \binrel spacing to \overset and \underset.
+        // \binrel spacing varies with (bin|rel|ord) of the atom in the argument.
+        // We'll do the same.
+        if (baseArg.type === "ordgroup") {
+            mclass = "m" + baseArg.value[0].type.replace("math", "");
+        } else {
+            mclass = "m" + baseArg.type.replace("math", "");
```
Comment (edemaine) [inline src/functions.js]: Is `baseArg.type` / `baseArg.value[0].type` guaranteed to be one of `mathbin`/`mathrel`,`mathord`?  What if there's an `ordgroup` in an `ordgroup`, for example?...  I'm not suggesting going any deeper, but I worry that we need additional checking here.

Comment (ronkok) [reply src/functions.js]: > What if there's an `ordgroup` in an `ordgroup`?

Then the result will have a CSS class `mordgroup`, that is, a class with no style rules attached. Clearly that is not ideal, but I didn't really want to do a recursive search.

In the current KaTeX release, `\overset` and `\underset` both assign `\mathop` spacing regardless of the argument type. Would it be okay if I searched for an a (bin|rel|ord) type in the top two levels, use it if found, but use `\mathop` for a default if the `ordgroup`s go deeper?

Comment (ronkok) [reply src/functions.js]: @edemaine I just made a commit that tries to resolve nested `ordgroups`. Thank you for the nudge. What I had before was sloppy.

Mind you, what I have now does seem particularly clean, but nothing better came to mind.

**Comment by ronkok** [comment]
Well, the arguments passed to this function are ParseNodes, not DomNodes, so `getTypeOfDomTree` is not helpful without a lot of pre-processing. That seems kind of wasteful. Let me suggest that the base of a `\overset` can fall into three categories:

| Category                           | Example                           | We want mclass to be: |
|:----------------------------|:----------------------------|:--------------------------|
| single symbol                    | \\overset ? =                   | baseArg.type              |
| group containing a single symbol | \\overset {?}{=}   | baseArg\[0].type          |
| everything else                  | \\overset ? {{a+b} \\div c} | `mord`                     |

If that's true, then we can get the needed information with:
```
		let atomType = "";
		if (baseArg.type === "ordgroup") {
			atomType = baseArg.value[0].type;
		} else {
			atomType = baseArg.type;
		}
		if (/bin|rel/.test(atomType)) {
			mclass = "m" + atomType;
		} else {
			mclass = "mord";
		}
```

Yes, what I've written is a bit of hack, but after looking at that LaTeX code, I don't feel too bad about that. And this code stays local and fast, without excursions into `buildHTML`. 

What do you think?

**Comment by edemaine** [comment]
@ronkok This seems almost perfect.  However, I believe there's some confusion about the cases, because `\overset{?}{=}` and `\overset ? =` should parse identically (and do -- I just checked).  So the arguments are *always* wrapped in `ordgroup`.

That said, I think the code you posted is exactly what we want.  This way we can handle a non-ordgroup value, just in case.  We don't want to go any deeper, though, so that `\overset{{?}}{{=}}` behaves as it does in LaTeX (I think) -- treating the arguments like an ord group.

**Comment by ronkok** [comment]
> some confusion about the cases

> the code you posted is exactly what we want.

Better lucky than good, eh?

I'll make the change and commit.

**Comment by edemaine** [comment]
Hmm, it occurs to me that, because this code is looking at the parse tree, it won't properly handle other parse tree nodes that end up acting like a `bin` or `rel`.  For example, an `\overset` within an `\overset` argument won't behave correctly with this code, but ought to with LaTeX.

One alternative would be to wait until HTML/MathML build time, and use the computed class.  This would be closer to LaTeX's behavior.  I'm not sure whether it's worth the added complexity... maybe at least a comment?

**Comment by ronkok** [comment]
That's reasonable. I added a comment.

**Comment by edemaine** [comment]
I'm good with this. @kevinbarabash did you want to look at it, or should we just merge?

**Comment by kevinbarabash** [comment]
@ronkok thanks for fixing this.  @edemaine thanks for reviewing.   The changes look reasonable.  Let's merge it!

File: src/functions.js [hunk 2]
```diff
@@ -70,26 +70,48 @@ defineFunction("mclass", [
...
+    const baseOp = new ParseNode("op", {
         type: "op",
         limits: true,
         alwaysHandleSupSub: true,
         symbol: false,
-        value: ordargument(mathAxisArg),
-    }, mathAxisArg.mode);
+        suppressBaseShift: context.funcName !== "\\stackrel",
+        value: ordargument(baseArg),
+    }, baseArg.mode);
```
Comment (ylemkimon) [inline src/functions.js]: `ParseNode`s are flow-typed by #1276:
https://github.com/Khan/KaTeX/blob/5a4aedd8827776acd737c993b9416c36302465f7/src/ParseNode.js#L66-L73

Comment (ronkok) [reply src/functions.js]: @ylemkimon Thank you for adding `suppressBaseShift` to the flow type enumeration. I was trying to think of a way to make `\underset` work without doing that. I was getting nowhere.

**Pull request merged**